components/psst/browser/core/psst_rule_registry_unittest.cc

```diff
@@ -0,0 +1,235 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/browser/core/psst_rule_registry.h"
+
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <utility>
+
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/functional/callback_forward.h"
+#include "base/memory/raw_ptr.h"
+#include "base/path_service.h"
+#include "base/test/mock_callback.h"
+#include "base/test/scoped_feature_list.h"
+#include "brave/components/constants/brave_paths.h"
+#include "brave/components/psst/browser/core/matched_rule.h"
+#include "brave/components/psst/browser/core/rule_data_reader.h"
+#include "brave/components/psst/common/features.h"
+#include "content/public/test/browser_task_environment.h"
+#include "testing/gmock/include/gmock/gmock.h"
+#include "testing/gtest/include/gtest/gtest.h"
+#include "url/gurl.h"
+
+namespace psst {
+
+namespace {
+// Test PSST rules file: brave/test/data/psst-component-data/psst.json
+constexpr size_t kTestPsstRulesCount = 3;
+constexpr char kPsstUserScriptName[] = "user.js";
+constexpr char kPsstPolicyScriptName[] = "policy.js";
+constexpr char kPsstJsonFileName[] = "psst.json";
+
+std::string ReadFile(const base::FilePath& file_path) {
+  std::string contents;
+  bool success = base::ReadFileToString(file_path, &contents);
+  if (!success || contents.empty()) {
+    VLOG(2) << "ReadFile: cannot " << "read file " << file_path;
+  }
+  return contents;
+}
+}  // namespace
+
+using OnRuleMatchedCallback = base::RepeatingCallback<void(
+    const std::optional<MatchedRule>& matched_rule)>;
+
+class PsstRuleRegistryUnitTest : public testing::Test {
+ public:
+  void SetUp() override {
+    scoped_feature_list_.InitAndEnableFeature(features::kEnablePsst);
+
+    base::FilePath test_data_dir(
+        base::PathService::CheckedGet(brave::DIR_TEST_DATA));
+    test_data_dir_base_ = test_data_dir.AppendASCII("psst-component-data");
+  }
+
+  base::FilePath GetTestDataDirBase() const { return test_data_dir_base_; }
+  base::FilePath GetScriptsTestDataDir() const {
+    return GetTestDataDirBase().Append(
+        base::FilePath::FromUTF8Unsafe("scripts"));
+  }
+
+ private:
+  content::BrowserTaskEnvironment task_environment_;
+  base::test::ScopedFeatureList scoped_feature_list_;
+  base::FilePath test_data_dir_base_;
+};
+
+TEST_F(PsstRuleRegistryUnitTest, LoadConcreteRule) {
+  PsstRuleRegistry::GetInstance()->ResetRuleRegistryForTest();
```

Comment:
Doing unit testing with the singleton method is a bit of a problem for this reason. In this case I think it's better to have a test friend that can instantiate `PsstRuleRegistry` directly. If we were going to use the singleton `ResetRuleRegistryForTest` should go in `SetUp`, but this seems very fragile in general because any new state that gets added in `PsstRuleRegistry` needs to be reset so this is really not a great pattern to use.